List security profiles applicable on a given target

jan-cerny opened the pull request:
This commit introduce 'oscapd-evaluate target-profiles' subcommand.
Similarly to 'oscapd-evaluate target-cpes', this will show a list
of available profiles to a scan target.

mpreisler commented on openscap_daemon/cli_helpers.py:
```diff
@@ -302,3 +302,27 @@ def summarize_standard_compliance_results(arf_source, result_list):
 
         result_list.append(result_json)
+
+def get_list_of_profiles(sds_filepath):
```
There already is a function for this. Please use that instead. It's called scrape_profiles or something like that.

jan-cerny replied:
@mpreisler  Aha OK, I was looking for something like that before I implemented this, but I must be blind. I will try to reuse that here.

jan-cerny commented:
I used `oscap_helpers.get_profile_choices_for_input()` instead.

mpreisler commented on openscap_daemon/oscap_helpers.py:
```diff
@@ -125,7 +125,7 @@ def scrape_profiles(tree, namespace, dest):
         )
 
-    ret[""] = "(default)"
+    ret["(default)"] = "Default profile"
```
??? This will break existing use-cases of this API.

jan-cerny replied:
Thanks for suggestion. This change isn't necessary. I'll remove it.

jan-cerny commented:
I have removed the commit with change of default profile. I have also fixed the std output for `oscapd-evaluate target-profiles`. I have rebased the pull request.

mpreisler commented on bin/oscapd-evaluate:
```diff
@@ -517,6 +529,20 @@ def main():
+        ssg_sds = config.get_ssg_sds(cpes)
+        print("Security profiles applicable on target " + args.target + ":")
+        profiles = oscap_helpers.get_profile_choices_for_input(ssg_sds, None)
+        for profile_id, title in profiles.items():
```
I wonder what happens if the SDS has multiple benchmarks.

jan-cerny commented:
@mpreisler That's a good question.

All profiles from all benchmarks will be printed.

The problem is that the XCCDF ID is not printed here. Moreover, `oscapd evaluate scan` has no way to choose benchmark within a datastream. Default benchmark is always chosen. 

Scan won't work for `xccdf_org.ssgproject.content_profile_pci-dss_centric` on RHEL7, even though this will be listed here.

I have encountered same behaviour of `oscapd-cli task-create`. It also allows me to create a task with profile `xccdf_org.ssgproject.content_profile_pci-dss_centric` on RHEL7 , but running the task errors out, because it can't find the profile in the datatstram.

jan-cerny commented:
@mpreisler I have further investigated the issue with multiple benchmarks in one SDS. I think about 2 solutions:

1) Use only the default benchmark everywhere, don't list profiles from other benchmarks. That's would be very easy to do. I don't like the solution, because we couldn't use the PCI-DSS centric profile from SSG or something else in future.

2) Support multiple benchmarks in one datastream. That would require to rewrite some of the existing methods, namely `get_profile_choices_for_input`, so that it doesn't return a dictionary [id:title], but rather a dictionary of dicitionaries. Also implement `--bechmark-id` in `oscapd-evaluate scan`. Also implement support for multiple benchmarks in interactive task creating.

What do you think about this?

jan-cerny commented:
Solution 1 is a 6-line patch. I can push the patch here as a temporary solution, and work on solution 2 in a separate pull request. The patch won't introduce a regression, because multiple benchmarks in a single datastream aren't supported now, as I mentioned before.

mpreisler commented:
@jan-cerny agreed

jan-cerny commented:
@mpreisler OK, I have pushed the temporary patch.

mpreisler commented on bin/oscapd-evaluate:
```diff
@@ -450,6 +450,18 @@ def main():
     )
+    target_profiles_parser = subparsers.add_parser(
+        "target-profiles",
+        help="Detect security profiles applicable on given target"
```
This is specifically for SCAP Security Guide and SSG only, right? Shouldn't the help text say that?

jan-cerny replied:
Yes, it is, it lists profiles for SSG. Thanks for suggestion. I will improve the help.

mpreisler commented:
ACK
